### PR TITLE
Draft: FilePath.ComponentView

### PR DESCRIPTION
**Pull request by milseman**
Draft: Add FilePath.ComponentView, a Bidi RRC of components
    
Adds in testing infrastructure and collectiony tests, as well
as ad-hoc testing.

### CODE REVIEW

File: Sources/System/FilePathComponents.swift [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+
+    // Explicitly add a trailing separator if
+    //   not at end and next character is not a separator
+    let atEnd = upperBound == endIndex
+    let trailingSeparator = !atEnd && !isSeparator(path.bytes[upperBound._storage])
+
+    // Explicitly add a preceding separator if
+    //   replacing front with absolute components (unless redundant by trailing separator),
+    //   preceding character is not a separator (which implies at the end)
+    let atStart = lowerBound == startIndex
```
Comment (milseman) [inline Sources/System/FilePathComponents.swift:279]: TODO: Account for `//` special case...

> Multiple successive <slash> characters are considered to be the same as one <slash>, except for the case of exactly two leading <slash> characters.

https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_271

Comment (milseman) [reply Sources/System/FilePathComponents.swift:279]: For POSIX, it seems to denote implementation defined path resolution:

> A pathname consisting of a single <slash> shall resolve to the root directory of the process. A null pathname shall not be successfully resolved. If a pathname begins with two successive <slash> characters, the first component following the leading <slash> characters may be interpreted in an implementation-defined manner, although more than two leading <slash> characters shall be treated as a single <slash> character.

https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_13

Comment (lorentey) [reply Sources/System/FilePathComponents.swift:279]: Luckily, all signs indicate that `//` is special cased on neither Darwin nor Linux, so there is no need for us to do anything about it. It's okay to treat `//` as a synonym of `/` for the purposes of canonicalization and component extraction.

**Comment by milseman** [comment]
This work is superseded by FilePath syntactic operations: https://github.com/apple/swift-system/pull/14

**Pull request closed without merging**